File: src/main/java/de/photon/aacadditionpro/util/visibility/PlayerInformationHider.java
package de.photon.aacadditionpro.util.visibility;

import com.comphenix.protocol.PacketType;
import com.comphenix.protocol.ProtocolLibrary;
import com.comphenix.protocol.events.ListenerPriority;
import com.comphenix.protocol.events.PacketAdapter;
import com.comphenix.protocol.events.PacketEvent;
import com.comphenix.protocol.events.PacketListener;
import com.google.common.collect.Multimap;
import com.google.common.collect.MultimapBuilder;
import de.photon.aacadditionpro.AACAdditionPro;
import de.photon.aacadditionpro.ServerVersion;
import lombok.val;
import org.bukkit.Bukkit;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.world.ChunkUnloadEvent;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;

abstract class PlayerInformationHider implements Listener
{
    private final PacketListener informationPacketListener;
    private final Multimap<Entity, Entity> hiddenFromPlayerMap;

    protected PlayerInformationHider(@NotNull PacketType... affectedPackets)
    {
        informationPacketListener = new PacketAdapter(AACAdditionPro.getInstance(), ListenerPriority.NORMAL, Set.of(affectedPackets))
        {
            @Override
            public void onPacketSending(final PacketEvent event)
            {
                if (event.isPlayerTemporary()) return;
                val entityID = event.getPacket().getIntegers().read(0);

                final boolean hidden;
                synchronized (hiddenFromPlayerMap) {
                    hidden = hiddenFromPlayerMap.get(event.getPlayer()).stream().map(Entity::getEntityId).anyMatch(i -> i.equals(entityID));
                }

                if (hidden) event.setCancelled(true);
            }
        };

        hiddenFromPlayerMap = MultimapBuilder.hashKeys(AACAdditionPro.SERVER_EXPECTED_PLAYERS).hashSetValues(AACAdditionPro.WORLD_EXPECTED_PLAYERS).build();
    }

    public void clear()
    {
        synchronized (hiddenFromPlayerMap) {
            hiddenFromPlayerMap.clear();
        }
    }

    public void registerListeners()
    {
        // Only start if the ServerVersion is supported
        if (ServerVersion.containsActiveServerVersion(this.getSupportedVersions())) {
            // Register events and packet listener
            AACAdditionPro.getInstance().registerListener(this);
            ProtocolLibrary.getProtocolManager().addPacketListener(this.informationPacketListener);
        }
    }

    public void unregisterListeners()
    {
        HandlerList.unregisterAll(this);
        ProtocolLibrary.getProtocolManager().removePacketListener(this.informationPacketListener);
    }

    protected Set<ServerVersion> getSupportedVersions()
    {
        return EnumSet.allOf(ServerVersion.class);
    }

    @EventHandler
    public void onEntityDeath(EntityDeathEvent event)
    {
        removeEntity(event.getEntity());
    }

    @EventHandler
    public void onChunkUnload(ChunkUnloadEvent event)
    {
        // Cache entities for performance reasons so the server doesn't need to load them again when the
        // task is executed.
        synchronized (hiddenFromPlayerMap) {
            for (final Entity entity : event.getChunk().getEntities()) removeEntity(entity);
        }
    }

    @EventHandler
    public void onQuit(PlayerQuitEvent event)
    {
        removeEntity(event.getPlayer());
    }

    /**
     * Remove the given entity from the underlying map.
     *
     * @param entity - the entity to remove.
     */
    private void removeEntity(Entity entity)
    {
        synchronized (hiddenFromPlayerMap) {
            hiddenFromPlayerMap.removeAll(entity);
            // Remove all the instances of entity from the values.
            //noinspection StatementWithEmptyBody
            while (hiddenFromPlayerMap.values().remove(entity)) ;
        }
    }

    /**
     * Hides a {@link Player} from another {@link Player}.
     */
    public void setHiddenEntities(@NotNull Player observer, @NotNull Collection<Entity> toHide)
    {
        Collection<Entity> oldEntities;
        synchronized (hiddenFromPlayerMap) {
            oldEntities = hiddenFromPlayerMap.replaceValues(observer, toHide);
        }

        // ProtocolManager check is needed to prevent errors.
        if (ProtocolLibrary.getProtocolManager() != null) {
            for (Entity entity : oldEntities) {
                if (!toHide.contains(entity)) updateEntity(observer, entity);
            }
        }

        this.onHide(observer, toHide);
    }

    public void revealAllEntities(@NotNull Player observer)
    {
        Collection<Entity> oldEntities;
        synchronized (hiddenFromPlayerMap) {
            oldEntities = hiddenFromPlayerMap.removeAll(observer);
        }

        for (Entity oldEntity : oldEntities) updateEntity(observer, oldEntity);
    }

    public void updateEntity(@NotNull Player observer, @NotNull Entity entity)
    {
        Bukkit.getScheduler().runTask(AACAdditionPro.getInstance(), () -> ProtocolLibrary.getProtocolManager().updateEntity(entity, List.of(observer)));
    }

    protected abstract void onHide(@NotNull Player observer, @NotNull Collection<Entity> toHide);
}

File: src/main/java/de/photon/aacadditionpro/util/visibility/PlayerInformationHider.kt
package de.photon.aacadditionpro.util.visibility

import com.comphenix.protocol.PacketType
import com.comphenix.protocol.ProtocolLibrary
import com.comphenix.protocol.events.ListenerPriority
import com.comphenix.protocol.events.PacketAdapter
import com.comphenix.protocol.events.PacketEvent
import com.comphenix.protocol.events.PacketListener
import com.google.common.collect.ImmutableList
import com.google.common.collect.ImmutableSet
import com.google.common.collect.Multimap
import com.google.common.collect.MultimapBuilder
import de.photon.aacadditionpro.AACAdditionPro
import de.photon.aacadditionpro.ServerVersion
import de.photon.aacadditionpro.user.data.Constants
import org.bukkit.Bukkit
import org.bukkit.entity.Entity
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDeathEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.event.world.ChunkUnloadEvent
import java.util.*

internal abstract class PlayerInformationHider protected constructor(vararg affectedPackets: PacketType) : Listener {
    private val informationPacketListener: PacketListener
    private val hiddenFromPlayerMap: Multimap<Int, Int>

    fun clear() {
        synchronized(hiddenFromPlayerMap) {
            hiddenFromPlayerMap.clear()
        }
    }

    fun registerListeners() {
        // Only start if the ServerVersion is supported
        if (ServerVersion.supportsActiveServerVersion(supportedVersions)) {
            // Register events and packet listener
            AACAdditionPro.getInstance().registerListener(this)
            ProtocolLibrary.getProtocolManager().addPacketListener(informationPacketListener)
        }
    }

    fun unregisterListeners() {
        HandlerList.unregisterAll(this)
        ProtocolLibrary.getProtocolManager().removePacketListener(informationPacketListener)
    }

    protected open val supportedVersions: Set<ServerVersion> get() = EnumSet.allOf(ServerVersion::class.java)

    @EventHandler
    fun onEntityDeath(event: EntityDeathEvent) {
        removeEntity(event.entity)
    }

    @EventHandler
    fun onChunkUnload(event: ChunkUnloadEvent) {
        // Cache entities for performance reasons so the server doesn't need to load them again when the
        // task is executed.
        synchronized(hiddenFromPlayerMap) {
            for (entity in event.chunk.entities) removeEntity(entity)
        }
    }

    @EventHandler
    fun onQuit(event: PlayerQuitEvent) {
        removeEntity(event.player)
    }

    /**
     * Remove the given entity from the underlying map.
     *
     * @param entity - the entity to remove.
     */
    @Suppress("ControlFlowWithEmptyBody")
    private fun removeEntity(entity: Entity) {
        synchronized(hiddenFromPlayerMap) {
            hiddenFromPlayerMap.removeAll(entity.entityId)
            // Remove all the instances of entity from the values.
            while (hiddenFromPlayerMap.values().remove(entity.entityId));
        }
    }

    /**
     * Hides a [Player] from another [Player].
     */
    fun hidePlayer(observer: Player?, playerToHide: Player?) {
        checkNotNull(observer) { "Tried to hide information from null observer." }
        checkNotNull(playerToHide) { "Tried to hide information of null player." }

        synchronized(hiddenFromPlayerMap) {
            hiddenFromPlayerMap.put(observer.entityId, playerToHide.entityId)
        }

        onHide(observer, playerToHide)
    }

    protected abstract fun onHide(observer: Player, playerToHide: Player)

    /**
     * Unhides a [Player] from another [Player].
     */
    fun revealPlayer(observer: Player?, playerToReveal: Player?) {
        checkNotNull(observer) { "Tried to reveal information to null observer." }
        checkNotNull(playerToReveal) { "Tried to reveal information of null player." }

        var hiddenBefore: Boolean
        synchronized(hiddenFromPlayerMap) {
            hiddenBefore = hiddenFromPlayerMap.remove(observer.entityId, playerToReveal.entityId)
        }

        // Resend packets
        if (ProtocolLibrary.getProtocolManager() != null && hiddenBefore) {
            Bukkit.getScheduler().runTask(AACAdditionPro.getInstance(), Runnable { ProtocolLibrary.getProtocolManager().updateEntity(playerToReveal, ImmutableList.of(observer)) })
        }
    }

    init {
        hiddenFromPlayerMap = MultimapBuilder.hashKeys(Constants.SERVER_EXPECTED_PLAYERS).hashSetValues(Constants.WORLD_EXPECTED_PLAYERS).build()

        informationPacketListener = object : PacketAdapter(AACAdditionPro.getInstance(), ListenerPriority.NORMAL, ImmutableSet.copyOf(affectedPackets)) {
            override fun onPacketSending(event: PacketEvent) {
                val entityID = event.packet.integers.read(0)

                if (!event.isPlayerTemporary) {
                    val hidden: Boolean
                    synchronized(hiddenFromPlayerMap) {
                        hidden = hiddenFromPlayerMap.containsEntry(event.player.entityId, entityID)
                    }
                    if (hidden) event.isCancelled = true
                }
            }
        }
    }
}